bindings/rust/bench/src/harness.rs

```diff
@@ -33,131 +89,266 @@ pub enum HandshakeType {
+        crypto_config: CryptoConfig,
+        handshake_type: HandshakeType,
+    ) -> Result<Self::Config, Box<dyn Error>>;
+
+    /// Make connection from existing config and buffer
+    fn new_from_config(
+        config: &Self::Config,
+        connected_buffer: ConnectedBuffer,
+    ) -> Result<Self, Box<dyn Error>>;
+
     /// Initialize buffers, configs, and connections (pre-handshake)
     fn new(
+        mode: Mode,
         crypto_config: CryptoConfig,
         handshake_type: HandshakeType,
         buffer: ConnectedBuffer,
-    ) -> Result<Self, Box<dyn Error>>;
+    ) -> Result<Self, Box<dyn Error>> {
+        Self::new_from_config(
+            &Self::make_config(mode, crypto_config, handshake_type)?,
+            buffer,
+        )
+    }
 
-    /// Run handshake on initialized connections
-    /// Returns error if handshake has already completed
+    /// Run one handshake step: receive msgs from other connection, process, and send new msgs
     fn handshake(&mut self) -> Result<(), Box<dyn Error>>;
 
-    /// Checks if handshake is finished for both client and server
     fn handshake_completed(&self) -> bool;
 
-    /// Get negotiated cipher suite
     fn get_negotiated_cipher_suite(&self) -> CipherSuite;
 
-    /// Get whether or negotiated version is TLS1.3
     fn negotiated_tls13(&self) -> bool;
 
-    /// Send application data from connection in harness pair
-    fn send(&mut self, sender: Mode, data: &[u8]) -> Result<(), Box<dyn Error>>;
+    /// Send application data to ConnectedBuffer
+    fn send(&mut self, data: &[u8]) -> Result<(), Box<dyn Error>>;
+
+    /// Read application data from ConnectedBuffer
+    fn recv(&mut self, data: &mut [u8]) -> Result<(), Box<dyn Error>>;
+
+    /// Shrink buffers owned by the connection
+    fn shrink_connection_buffers(&mut self);
+
+    /// Clear and shrink buffers used for IO with another connection
+    fn shrink_connected_buffer(&mut self);
 
-    /// Receive application data sent to connection in harness pair
-    fn recv(&mut self, receiver: Mode, data: &mut [u8]) -> Result<(), Box<dyn Error>>;
+    /// Get reference to internal connected buffer
+    fn connected_buffer(&self) -> &ConnectedBuffer;
+}
+
+pub struct TlsConnPair<C: TlsConnection, S: TlsConnection> {
+    client: C,
+    server: S,
+}
+
+impl<C: TlsConnection, S: TlsConnection> Default for TlsConnPair<C, S> {
+    fn default() -> Self {
+        Self::new(Default::default(), Default::default(), Default::default()).unwrap()
+    }
+}
+
+impl<C: TlsConnection, S: TlsConnection> TlsConnPair<C, S> {
+    /// Wrap two TlsConnections into a TlsConnPair
+    pub fn wrap(client: C, server: S) -> Self {
+        assert!(
+            client.connected_buffer() == &server.connected_buffer().clone_inverse(),
+            "connected buffers don't match"
+        );
+        Self { client, server }
+    }
+
+    /// Take back ownership of individual connections in the TlsConnPair
+    pub fn unwrap(self) -> (C, S) {
```

Comment:
`unwrap` is confusing naming given that `unwrap` already has a defined use in the Rust ecosystem. Perhaps `consume`?

Reply:
`consume` makes it sound like that the connections are being used up and turned invalid. Would `split()` or `take_ownership()` work? I'm leaning towards `split()` personally, so I'll commit that change for now.

Reply:
Alternately, is ownership actually important? if not `inner() -> (&C, &S)` is a pretty common pattern

Reply:
This is useful for my memory benching experiments where a connection is handshaked but only the client or only the server is stored and not dropped.